### ISSUE TITLE
Support Prometheus histograms in the heatmap panel

### ISSUE THREAD

**Issue by bergquist** [grafana/grafana#10009]
There are three things that need to be solved to support Prometheus histograms in the heatmap panel.

### Transform the histogram result into the format expected by the heatmap panel
Prometheus stores the values of each bucket incrementally. So we have to decrement each bucket in the Prometheus data source before sending it to the heatmap panel. I started working on that in a branch but it's out of sync with master. The transform function can be found at https://github.com/bergquist/grafana/blob/prom_heatmap/public/app/plugins/datasource/prometheus/result_transformer.ts#L21 and one test for the function can be found at https://github.com/bergquist/grafana/blob/prom_heatmap/public/app/plugins/datasource/prometheus/specs/result_transformer.ts#L17 Those can be used to get started quickly. 

### We need to handle Prometheus `+Inf` serie. 
In Prometheus, each histogram contains a serie named `+Inf` which is used whenever a value does not find in any other predefined bucket. This serie can help you decide if you should are more buckets or not. I'm not sure how to visualize this serie, if at all. If we decide to not show it we should remove it from the response when transforming the result into heat map format. 

### Exponential buckets 
Using the helper method in the Prometheus client libs you usually choose between Linear and exponential buckets. Would be great if the heat map panel could support this as well log base scales.

**Comment by alexanderzobnin** [comment]
@bergquist I'm continuing work on this, rebased onto master (https://github.com/alexanderzobnin/grafana/tree/prom-heatmap)

**Comment by discordianfish** [comment]
That'd be awesome. Feel free to ping me whenever you need help/input from the prometheus side of things!

**Comment by neopaf** [comment]
Alexander, we're all praying for you.

**Comment by bergquist** [comment]
Closed by https://github.com/grafana/grafana/pull/11087

**Comment by bergquist** [comment]
This feature needs better docs.

**Comment by alexanderzobnin** [comment]
Could you create an issue for that, @bergquist ?

**Comment by alexanderzobnin** [comment]
I agree.

**Comment by neopaf** [comment]
If important, other tabs:
![2018-03-26 15 55 08](https://user-images.githubusercontent.com/4362564/37907893-17fc6cd2-310f-11e8-9721-870b2febb3c5.png)

![2018-03-26 15 55 13](https://user-images.githubusercontent.com/4362564/37907904-1a6cadba-310f-11e8-8b4e-e451d384444c.png)

**Comment by F30** [comment]
@neopaf What version of Grafana are you using? This feature just got merged 13 days ago, afaik it is not in a released version yet.

**Comment by neopaf** [comment]
@F30 5.1.0-13147pre1 (from Nightly builds).
I expected it to have the master merge (not sure if my assumptions are on solid ground)

**Comment by torkelo** [comment]
Think you need to use the quantile prometheus function

**Comment by discordianfish** [comment]
Also don't confuse summary with histograms. I haven't looked closer but grafana probably only supports histograms and what you have looks like a summary: https://prometheus.io/docs/practices/histograms/

**Comment by neopaf** [comment]
@torkelo @discordianfish 
Friends, I feel quantile/histograms are not related to my original understanding of this task.

On client side I observe events straight forwardly:
```scala
	private val histogram = Histogram.build.exponentialBuckets(0.0001, 10, 5).name("PMF").labelNames("entry", "stage", "status").help("In seconds.").register

	def observe(entry: String, stage: String, status: String, timer: SimpleTimer): Unit = {
		histogram.labels(entry, stage, status).observe(timer.elapsedSeconds)
```

And I get direct pre-bucketed values that need no estimates.
I thought, current task is designed to visualise precisely that information.

Anybody feels this change is in available nightly builds?

**Comment by neopaf** [comment]
But I did try this:
```
histogram_quantile(0.95, sum(PMF_bucket{entry="POST /gc/metaProfile", stage="mergeTwoServiceCubes", job="pmf", status="success"}) by (le))
```

And got this in Prometheus:
![2018-03-26 16 54 16](https://user-images.githubusercontent.com/4362564/37910731-ed19f324-3116-11e8-9698-a50487f2c544.png)
(clearly by bucket data is already lost)

And to-be-expected strange result in grafana:
![2018-03-26 16 55 20](https://user-images.githubusercontent.com/4362564/37910796-0850dc84-3117-11e8-9c08-2bbe3a49677d.png)

**Comment by neopaf** [comment]
But I did try this:
```
histogram_quantile(0.95, sum(rate(PMF_bucket{entry="POST /gc/metaProfile", stage="mergeTwoServiceCubes", job="pmf", status="success"}[1m])) by (le))
```
And got in Prometheus:
![2018-03-26 17 01 12](https://user-images.githubusercontent.com/4362564/37910943-704f4500-3117-11e8-97b9-c4805aacc2ba.png)
(I feel that information is already lost on this step)

In Grafana:
![2018-03-26 17 01 36](https://user-images.githubusercontent.com/4362564/37910959-7d8f7cd0-3117-11e8-8b55-52f8bcdc1623.png)

With strange bucket names (not exponential, not as created originally).

I truly don't understand what's going on with this function.
But my original understanding is that it is not needed.
My data is already pre-bucketed.

**Comment by oseiberts11** [comment]
@neopaf: I don't think you should use the histogram_quantile function here. That function really throws away a lot of data from the histogram; the outcome is definitely no longer a histogram so trying to show it as one would indeed show nonsense.

**Comment by bergquist** [comment]
As said earlier, avoid the `histrogram_quantile` function for visualization. Instead use something like this: 
`sum(rate(PMF_bucket{entry="POST /gc/metaProfile", stage="mergeTwoServiceCubes", job="pmf", status="success"}[1m])) by (le)`

and set `Legend format` to `{{le}}`

**Comment by neopaf** [comment]
Thanks, Carl.
Thanks, Alexander for your efforts and latest commit to docs https://github.com/alexanderzobnin/grafana/commit/b97cede054f22356405a55228bdea6f353e4d8ff

It works for me:

![2018-03-27 10 57 52](https://user-images.githubusercontent.com/4362564/37954290-09d943ec-31ae-11e8-846e-829410c41d17.png)

JSON of panel, for history:
[PAF heat map experiment-1522137445922.json.txt](https://github.com/grafana/grafana/files/1850949/PAF.heat.map.experiment-1522137445922.json.txt)

Prometheus view:
![2018-03-27 10 58 09](https://user-images.githubusercontent.com/4362564/37954353-34e56098-31ae-11e8-9e56-c761b1362aec.png)

I'm locally-happy :)
Thanks again!

**Comment by neopaf** [comment]
I feel key element was "format as Heatmap".
Maybe that should be a default when creating a Heatmap panel?

**Comment by alexanderzobnin** [comment]
@neopaf it's a default format for heatmap :)  not sure problem was in the format because it only converts histogram buckets from cumulative to regular. I think it's due to wrong label format, but not sure.

**Comment by neopaf** [comment]
@alexanderzobnin could be. I'm super happy. Thanks again!
(большое спасибо!)
